src/sentry/tagstore/v2/models/tagkey.py

```diff
@@ -91,6 +91,88 @@ def get_or_create(cls, project_id, environment_id, key, **kwargs):
 
         return rv, created
 
+    @classmethod
+    def get_or_create_bulk(cls, project_id, environment_id, keys):
+        # Attempt to create a bunch of models in one big batch with as few
+        # queries and cache calls as possible.
+        # In best case, this is all done in 1 cache get.
+        # In ideal case, we'll do 3 queries total instead of N.
+        # Absolute worst case, we still just do O(n) queries, but this should be rare.
+        key_to_model = {key: None for key in keys}
+        remaining_keys = set(keys)
+
+        # First attempt to hit from cache, which in theory is the hot case
+        cache_key_to_key = {cls.get_cache_key(project_id, environment_id, key): key for key in keys}
+        cache_key_to_models = cache.get_many(cache_key_to_key.keys())
+        for model in cache_key_to_models.values():
+            key_to_model[model.key] = model
+            remaining_keys.remove(model.key)
+
+        if not remaining_keys:
+            # 100% cache hit on all items, good work team
+            return key_to_model
+
+        # If we have some misses, we want to first check if
+        # all of the misses actually exist in the database
+        # already in one bulk query.
+        to_cache = {}
+        for model in cls.objects.filter(
+            project_id=project_id,
+            environment_id=environment_id,
+            key__in=remaining_keys,
+        ):
+            key_to_model[model.key] = to_cache[cls.get_cache_key(
+                project_id, environment_id, model.key)] = model
+            remaining_keys.remove(model.key)
+
+        # If we have found them all, cache all these misses
+        # and return all the hits.
+        if not remaining_keys:
+            cache.set_many(to_cache, 3600)
+            return key_to_model
+
+        # At this point, we need to create all of our keys, since they
+        # don't exist in cache or the database.
+
+        # First attempt to create them all in one bulk query
+        try:
+            with transaction.atomic():
+                cls.objects.bulk_create([
+                    cls(
+                        project_id=project_id,
+                        environment_id=environment_id,
+                        key=key,
+                    )
+                    for key in remaining_keys
+                ])
+        except IntegrityError:
+            pass
```

Comment:
shouldnt this raise an error or do something else other than pretend it created the rows?

Reply:
It hits on L169 and that's the fallback.

It's slightly disjointed because of the comment on L164-L165.